geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/cache/ThreadLocalCache.scala

```diff
@@ -0,0 +1,146 @@
+/***********************************************************************
+ * Copyright (c) 2013-2020 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.utils.cache
+
+import java.io.Closeable
+import java.lang.ref.WeakReference
+import java.util.concurrent._
+
+import com.github.benmanes.caffeine.cache.{Cache, Caffeine}
+import com.google.common.util.concurrent.MoreExecutors
+
+import scala.concurrent.duration.Duration
+
+/**
+ * Creates a per-thread cache of values with a timed expiration.
+ *
+ * Map operations will only affect/reflect the state of the current thread. Additional methods `globalIterator`
+ * and `estimatedGlobalSize` are provided for global views across all threads, which generally should only be used
+ * for debugging.
+ *
+ * Since caches are only cleaned up when accessed, uses an asynchronous thread to
+ * actively clean up any orphaned thread local values
+ *
+ * @tparam K key type
+ * @tparam V value type
+ */
+class ThreadLocalCache[K <: AnyRef, V <: AnyRef](
+    expiry: Duration,
+    executor: ScheduledExecutorService = ThreadLocalCache.executor
+  ) extends scala.collection.mutable.Map[K, V]
+    with scala.collection.mutable.MapLike[K, V, ThreadLocalCache[K, V]]
+    with Runnable
+    with Closeable {
```

Comment:
This is a lovely collection of interfaces to implement.  Maybe it should be serializable too just for kicks.

Reply:
yeah... i could make it private and just expose the Map interface to end users

Reply:
Nah, that's fine.  Just noting the interfaces; fun class!